docs/architecture/grpc-for-wcf-developers/client-libraries.md

```diff
@@ -1,11 +1,149 @@
+
+### IObservable
+
+The `IObservable<T>` interface is the "reactive" inverse of `IEnumerable<T>`. Rather than pulling items from a stream, the reactive approach lets the stream push items to a subscriber. This is very similar to gRPC streams, and it's easy to wrap an `IObservable<T>` around an `IAsyncStreamReader<T>`.
+
+This code is longer than the `IAsyncEnumerable<T>` code because C# doesn't have built-in support for working with observables, so the implementation class has to be created manually. It's a generic class, though, so a single implementation will work across all types.
+
+```csharp
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.Core
+{
+    public class GrpcStreamObservable<T> : IObservable<T>
+    {
+        private readonly IAsyncStreamReader<T> _reader;
+        private readonly CancellationToken _token;
+        private int _used;
+
+        public Observable(IAsyncStreamReader<T> reader, CancellationToken token = default)
+        {
+            _reader = reader;
+            _token = token;
+            _used = 0;
+        }
+
+        public IDisposable Subscribe(IObserver<T> observer) =>
+            Interlocked.Exchange(ref _used, 1) == 0
+                ? new GrpcStreamSubscription(_reader, observer, _token)
+                : throw new InvalidOperationException("Subscribe can only be called once.");
+
+    }
+}
+```
+
+> [!IMPORTANT]
+> This observable implementation only allows the `Subscribe` method to be called once, as having multiple subscribers trying to read from the stream would result in chaos. There are operators such as `Replay` in the [System.Reactive.Linq](https://www.nuget.org/packages/System.Reactive.Linq) that enable buffering and repeatable sharing of observables, which can be used with this implementation.
+
+The `GrpcStreamSubscription` class handles the enumeration of the `IAsyncStreamReader`.
+
+```csharp
+public class GrpcStreamSubscription : IDisposable
+{
+    private readonly Task _task;
+    private readonly CancellationTokenSource _tokenSource;
+    private bool _completed;
+
+    public Subscription(IAsyncStreamReader<T> reader, IObserver<T> observer, CancellationToken token)
+    {
+        Debug.Assert(reader != null);
+        Debug.Assert(observer != null);
+        _tokenSource = new CancellationTokenSource();
+        token.Register(_tokenSource.Cancel);
+        _task = Run(reader, observer, _tokenSource.Token);
+    }
+
+    private async Task Run(IAsyncStreamReader<T> reader, IObserver<T> observer, CancellationToken token)
+    {
+        while (!token.IsCancellationRequested)
+        {
+            try
+            {
+                if (!await reader.MoveNext(token)) break;
+            }
+            catch (RpcException e) when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
```

Comment:
```suggestion
            catch (RpcException e) when (e.StatusCode == Grpc.Core.StatusCode.Cancelled)
```

I assume cancelled is what you meant. I don't see why someone would ever return not found error from a method. To indicate there are no more items they would let the call end on the server.